docs/design/sum_types.md

```diff
@@ -0,0 +1,200 @@
+
+```carbon
+class Optional(T:! Type) {
+  // Factory functions
+  fn Some(value: T) -> Self;
+  let None:! Self;
+
+  private var has_value: bool;
+  private var value: T;
+
+  interface MatchContinuation {
+    let template ReturnType:! Type;
+    fn Some[addr me: Self*](value: T) -> ReturnType;
+    fn None[addr me: Self*]() -> ReturnType;
+  }
+
+  external impl as Match(MatchContinuation) {
+    fn Op[me: Self, Continuation:! MatchContinuation](
+        continuation: Continuation*) -> Continuation.ReturnType {
+      if (me.has_value) {
+        return continuation->Some(me.value);
+      } else {
+        return continuation->None();
+      }
+    }
+  }
+
+  // Operations like destruction, copying, assignment, and comparison are
+  // omitted for brevity.
+}
+```
+
+In this code, `Optional` makes itself available for use in pattern matching by
+declaring that it implements the `Match` interface. `Match` takes an interface
+argument, called the _continuation interface_, which specifies the set of
+possible alternatives by declaring a method for each one. In this case, we pass
+`Optional.MatchContinuation` as the continuation interface.
+
+When compiling a `match` statement, the compiler checks that the type being
+matched implements `Match(C)` for some continuation interface `C`. Then, it
+notionally transforms the `match` body into a class that implements `C`, with
+one method for each `case`, and then passes that to `Match.Op` on the object
+being matched. For example, the `match` statement shown earlier might be
+transformed into:
+
+```carbon
+class __MatchStatementImpl {
+  fn Make() -> Self { return {}; }
+
+  impl as Match(Optional.MatchContinuation) where .ReturnType = () {
+    fn Some(the_value: i32) {
+      Print(the_value);
+    }
+    fn None() {
+      Print("None");
+    }
+  }
+}
+
+my_opt.(Match.Op)(__MatchStatementImpl.Make());
+```
+
+Thus, a `match` statement works by invoking the sum type's `Match.Op` method,
+which is responsible for determining which alternative the sum object
+represents, and then invoking the compiler-supplied continuation that
+corresponds to that alternative. In order for this scheme to work, `Match.Op` is
+required to invoke exactly one method of `MatchContinuation`, and to do so
+exactly once.
+
+Notice that the names `Some` and `None` are defined twice, once as factory
+functions/constant members of `Optional` and once as methods of
+`MatchContinuation`, with the same parameter types in each case. The two
+effectively act as inverses of each other: the `Optional` members are used to
+create `Optional` values, and the `MatchContinuation` methods are used to report
+the parameter values that would create a given `Optional`. This mirroring
+between expression and pattern syntax is ultimately a design choice by the type
+author; there is no language-level requirement that the alternatives correspond
+to the factory functions, but it is **strongly** recommended.
```

Comment:
Is it worth pointing out here the consequences for not following this recommendation: specifically that the behavior of expression patterns and proper patterns diverge?